src/solve.jl

```diff
@@ -17,71 +17,493 @@ function DiffEqBase.__solve(prob::BVProblem, alg::Shooting; kwargs...)
+    elseif info == -4
+        retcode = ReturnCode.Terminated
+    else
+        retcode = ReturnCode.Failure
+    end
+   return DiffEqBase.build_solution(prob, alg, x, S.y; opt.retcode)
+end
+
+"""
+    interp_eval(mesh, Y, t, k_discrete, k_interp)
+
+After we construct an interpolant, we use interp_eval to evaluate it.
+"""
+function interp_eval(mesh, Y, alg, t, dt, len, s, s_star, k_discrete, k_interp)
+    # EXPORTS: z, z_prime
+    i = interval(mesh, t)
+    hi = mesh[i + 1] - mesh[i]
+    tau = (t - mesh[i]) / hi
+    weights, weights_prime = interp_weights(tau, alg)
+    z, z_prime = sum_stages(weights, weights_prime, k_discrete, k_interp, len, dt, Y, s,
+                            s_star)
+    return z, z_prime
+end
+
+function interval(mesh, t)
+    ind = findfirst(x -> x > t, mesh)
+    i::Int64 = copy(ind)
+    return i
+end
+
+"""
+    mesh_selector(mesh_current, defect, tol, n, len, alg)
+
+Generate new mesh based on the defect.
+"""
+function mesh_selector(mesh_current::Vector, defect, tol, n::Int64, len::Int64,
+                       alg::Union{GeneralMIRK, MIRK})
+    #exports: mesh_new, Nsub_star, info
+
+    #TODO: Need users to manually specify, here, we set it as 3000 by default.
+    MxNsub = 3000
+
+    safety_factor = 1.3
+    rho = 1.0 # Set rho=1 means mesh distribution will take place everytime.
+    upper_new_mesh = 4.0
+    lower_new_mesh = 0.5
+    r1 = 0.0
+    r2 = 0.0
+    Nsub_star = 0
+    info = 0
+    p = alg_order(alg)
+    s_hat = zeros(Float64, n)
+    for i in 1:n
+        h = mesh_current[i + 1] - mesh_current[i]
+        norm = abs(defect[i, idamax(defect[i, :])])
+        s_hat[i] = (norm / tol)^(1.0 / (p + 1)) / h
+        if s_hat[i] * h > r1
+            r1 = s_hat[i] * h
+        end
+        r2 = r2 + s_hat[i] * h
+    end
+    r3 = r2 / n
+    n_predict::Int64 = round(Int, (safety_factor * r2) + 1)
+    if abs((n_predict - n) / n) < 0.1
+        n_predict = round(Int, 1.1 * n)
+    end
+
+    if r1 <= rho * r3
+        Nsub_star::Int64 = 2 * n
+        if Nsub_star > MxNsub # Need to determine the too large threshold
+            println("New mesh would be too large")
+            info = -1
         else
-            general_eval_bc_residual!(S)
+            println("Half the current mesh")
+            mesh_new = half_mesh(mesh_current, n)
         end
-        flatten_vector!(resid, S.residual)
-        reorder!(resid)
-        return nothing
+    else
+        Nsub_star = copy(n_predict)
+        if Nsub_star > upper_new_mesh * n
+            Nsub_star = upper_new_mesh * n
+        end
+        if Nsub_star < lower_new_mesh * n
+            Nsub_star = lower_new_mesh * n
+        end
+        if Nsub_star > MxNsub
+            # Mesh redistribution fails
+            println("New mesh would be too large")
+            info = -1
+        else
+            println("Mesh redistributing")
+            mesh_new = redistribute(mesh_current, n, Nsub_star, s_hat)
```

Comment:
I don't undersatnd why one would redistribute. Wouldn't this make it so you aren't clustering the points where the error is the most anymore?

Reply:
From my understanding, after we redistribute the mesh and by using the new mesh to generate a new discrete solution, the new solution would have a smaller error.